From dfd318f2925aaf432b0a42b18dc76fc41bdabbd6 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 06 Aug 2020 12:31:00 +0300
Subject: [PATCH] [SW-16674] recovery: add bss section

This patch adds bss section for recovery image which will be
needed for adding module cmd support. Since recovery image is
only used on ARC, remove all the ifdef from startup file.

Change-Id: I2f3b23482ffab5086247adc6021a8dc0065b2766
---

diff --git a/goya2_arc_gnu/recovery.lds.S b/goya2_arc_gnu/recovery.lds.S
index fe20ea8..7873942 100644
--- a/goya2_arc_gnu/recovery.lds.S
+++ b/goya2_arc_gnu/recovery.lds.S
@@ -21,6 +21,13 @@
 		* (.text*)
 	} >.sram
 
+	.bss : {
+		. = ALIGN(4);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
 	.stack : {
 		. = ALIGN(4);
 		__stack = . ;
diff --git a/recovery/crt0_recovery.S b/recovery/crt0_recovery.S
index c948e58..acffc4a 100644
--- a/recovery/crt0_recovery.S
+++ b/recovery/crt0_recovery.S
@@ -11,17 +11,19 @@
 .globl	_start
 _start:
 
-#ifdef __arc__
+
 /* lf2flt: to avoid compiler (buildroot wrapper) warning during linkage */
 .global	lf2flt
 lf2flt:
-#endif
 
-#ifdef __arc__
-	push_s	blink
-#endif
+	mov	sp, __stack_top
+	mov	fp, sp
+
+	/* Clear BSS */
+	mov	r5, __bss_start
+	sub	r6, __bss_end, r5
+	lsr.f	lp_count, r6, 2
+	lpnz	1f
+	st.ab   0, [r5, 4]
+1:
 	bl	recovery_main	/* Set LR (x30) */
-#ifdef __arc__
-	pop_s	blink
-	j_s	[blink]
-#endif
